### ISSUE TITLE
Move TypeScript definitions to their own repository

### ISSUE THREAD

**Issue by Zalastax** [processing/p5.js#2571]
We just made a release, and the files on GitHub does not contain any TypeScript definitions. p5.zip doesn't have them either. They are, however, available in the [npm package](https://yarnpkg.com/en/package/p5?files).
Is there any way to test the release script properly, without accidentally releasing to npm/GitHub? When I made the modifications to the release script I tested it as best as I could, but it was not enough.

**Comment by Spongman** [comment]
i noticed they are missing from cdnjs, also: https://cdnjs.com/libraries/p5.js

**Comment by lmccart** [comment]
I can add them to github, but my preference is not to add them to the p5.zip to avoid confusion for beginners with many files in the folder.

**Comment by Zalastax** [comment]
Alright. Does that mean that you don't use the automated scripts when making a release? The scripts were modified to add the definition files to both the zip and GitHub, which is not what happened.

**Comment by lmccart** [comment]
I do use the automated scripts, I thought that the files were created erroneously.
I am wondering if the typescript should actually be a separate project rather than coming out of this build. It's feeling like too much for me to maintain properly. What are your thoughts, is this possible?

**Comment by Zalastax** [comment]
I'm sorry that it has increased the burden. It should be possible to maintain the definitions in another repo. It's more complicated, but the responsibility would be split to another maintainer.

The release scripts seem to be working then, so from my side everything is well. I don't think there will be many more changes to the TypeScript generator, just keep writing good and well-formatted documentation. If you want to wait on making a split to a repository for TypeScript, we can just change it so that the zip doesn't contain the definitions.

**Comment by lmccart** [comment]
@Zalastax yes I think this would be much preferred and better for the overall sustainability of the library. I don't really know anything about typescript so it makes it really difficult for me to tell what should be happening and how to debug when things aren't working. is migrating this to a separate repo something you'd be interested in taking on?

**Comment by Spongman** [comment]
I think it would be a shame to remove @toolness's tool from the build process, especially as it provides another really good source of validation for the doc comments. i was going to suggest that we even start having it throw errors during build to help ensure further that the doc comments are consistent during precommit. another repo would certainly work, but the lag between breakage and fix for issues like those would be increased. maybe that's not really a priority...

**Comment by Zalastax** [comment]
I think it would be a shame to remove it as well, but it increases complexity so we might have to maintain it elsewhere for now. My personal preference would actually be for p5 to go ES6 modules, then slowly transition to TypeScript, but that's really not up to me.
I won't have time to lead a move to another repo for half a year, but I could provide some help. A new repo would mean that we have to publish to DefinitelyTyped instead, right?
The TypeScript pipeline has found many problems with the documentation, so it would be great to keep that continuous feedback intact. Setting up continuous integration, following commits to p5, and opening issues/PRs here is probably enough.

**Comment by sakshamsaxena** [comment]
> Is there any way to test the release script properly, without accidentally releasing to npm/GitHub?

I think I should probably look into it, as a separate issue perhaps? Meanwhile, @Zalastax would you mind updating the title so this can grab attention of a possible person willing up to work on the maintenance you mention ?

**Comment by lmccart** [comment]
Picking this back up.. toward the goal of making this library more maintainable, I am strongly in favor of removing typescript completely from this repo and build process. We will transition at some point to ES6 and we could perhaps revisit it then. @Zalastax mentioned the possibility of moving this to another repo. Is anyone able to do this? Otherwise I think we will just need to remove / archive what's in place currently.

**Comment by toolness** [comment]
Er... sorry I have been absent from this discussion for so long!  Out of curiosity, has anyone brought up the possibility of having TypeScript infer typings from p5's existing JSDoc comments?

I couldn't look into this when I did my processing fellowship in 2016 because the feature wasn't actually in TypeScript yet: it was only added in a relatively recent version of the tool, I think at some point in 2017.  The TLDR is that TypeScript can actually parse JSDoc comments now, so you can get a lot of its benefits without even having to use TypeScript.

I'm happy to help look into this if folks want.  Sorry I have been absent!

**Comment by Zalastax** [comment]
@toolness I don't think TypeScript is compatible with YUIDoc, unfortunately.
I'm not sure how big the difference is, but migrating to JSDoc is probably a big task.
For now, it seems to me like the best path forward is to migrate the definitions generator to its own repository.
We can revisit this topic when ES6 is in place.
I would not be surprised if the path becomes ES6+YUIDoc->ES6+JSDoc->TS.

**Comment by toolness** [comment]
Oh shoot, my bad--I totally forgot that p5 uses YUIDoc and not JSDoc!  Hmmmm...

**Comment by lmccart** [comment]
hi @Zalastax that would be great, thank you! the repos in the processing organization represent ones that we have committed to maintaining (as best we can). since the reason for removing typescript from this repo is a lack of resources to maintain it, it would make sense to host this repo somewhere other than the processing org. you are welcome to set up in within your own set of repositories and we'd be happy to create easy links out to there from everywhere it makes sense.

**Comment by kjhollen** [comment]
Updated title here to reflect pending work items. Thanks!

**Comment by Zalastax** [comment]
I've set it all up!
You can find the [migrated generator](https://github.com/p5-types/p5.ts/) in the new [p5-types](https://github.com/p5-types) organization.
If we want generators for flow or Reason, they could live in the same organization.

In my migration I fixed some things in the generator, so I found a few more misses in the documentation.
Expect a Pull Request, fixing those, soon.

The next step is to remove files from this repository.
After the documentation is fixed we should also be ready to publish the typings to DefinetelyTyped.

**Comment by Zalastax** [comment]
To complete my work I am blocked on: #2954, #3010, #3011.
When they are merged I can create a PR to DefinetelyTyped and after that update our README.md to point people towards the then existing `@types/p5.js` package.

**Comment by Zalastax** [comment]
Everything is set up now and [p5.ts](https://github.com/p5-types/p5.ts) is able to generate typings that work well.
I would really need some help on how to import p5 in global mode correctly from TypeScript, but let's have that discussion over there instead.

**Comment by lmccart** [comment]
this is great, thank you @Zalastax! we might want to think about where to link to this from within the p5.js repo so people can find it more easily. maybe from the wiki somewhere? or the readme even?
